BugFix: try and prevent crashing during highlighting

SlySven opened the pull request:
It seems that one of the tests in `(void) TTextEdit::highlightSelection()` was being done too late - with the result that out of range indexing was being attempted on the `(std::deque<std::deque<TChar>) TBuffer::buffer` member. Specifically the test for `currentY >= totaly` should have been done before `currentY` was used to select one from the outer `std::deque` container so as to establish the size of it (to be stored as `maxX`).

The revised code replaces both the highlight and un-highlight code nested loops with components that are more conventionally used, i.e. with the test being based on a less than condition only with the limit being determined within the top parts of the for loops and no additional tests to break from the loop within the body thereof...

This PR was generated as a result of a crash report given by 2020/08/27 20:44 UTC in the Discord "**[#mudlet-development](https://discord.com/invite/bfvgTDz)**" channel by **Necerursh#2156** - other similar crashes may also be related to this issue - which is why I am marking this as a **high** priority bug/bugfix.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Cherry-pick: Add debugging echoes for copy crashing (#4000)

Added contents of PR #4000 (as at: b5459957b349a90b04b4757b6629b59d05011646 )
at Vadim's request...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: try and prevent crashing during highlighting

It seems that one of the tests in `(void) TTextEdit::highlightSelection()`
was being done too late - with the result that out of range indexing was
being attempted on the `(std::deque<std::deque<TChar>) TBuffer::buffer`
member. Specifically the test for `currentY >= totaly` should have been
done before `currentY` was used to select one from the outer `std::deque`
container so as to establish the size of it (to be stored as `maxX`).

The revised code replaces both the highlight and un-highlight code nested
loops with components that are more conventionally used, i.e. with the
test being based on a less than condition only with the limit being
determined within the top parts of the for loops and no additional tests
to break from the loop within the body thereof...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/TTextEdit.cpp:
```diff
@@ -645,28 +645,27 @@ void TTextEdit::highlightSelection()
-            }
+    // mpA represents the first (zero-based) line/row (y) and position/column
+    // (x) that IS to be selected, mpB represents the last (zero-based) line and
+    // column that IS to be selected - which means that in the traditional 'for'
```
This part of the comment about mpA and mpB duplicates the (simpler) text that's already above the function definition, and it's not ideal to describe the same thing in two different ways. Could you fix this?

SlySven committed:
BugFix: fix off-by-one errors in TTextEdit::getSelectedText(...)

The previous code did not allow for the fact that the actual end of line
and the end of the lines numbers are one less that the
`std::deque<T>::size()` method returns.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Thanks a lot for fixing this :+1: 

I find the new logic tough to read, but it seems to do the job.

vadi2 commented:
A crash still happened on 2b0e61c:

```
Thread 1 (Thread 0x7ffff0116900 (LWP 82177)):
#0  0x00000000008afd23 in TTextEdit::getSelectedText(QChar const&) ()
#1  0x00000000008aff7f in TTextEdit::slot_copySelectionToClipboard() ()
#2  0x00000000007a945e in TCommandLine::event(QEvent*) ()
```

SlySven commented:
> A crash still happened on 2b0e61c:

Was that on your own system perchance? Any more debug information available...?

vadi2 commented:
It's from the same user

vadi2 commented:
It has this debug information:

```
1 selection, buffer size 9090 startline 9364 endline - startline+1 -274
```

Then it crashed, so it didn't get too far past line 1435. That negative value doesn't look too good...

vadi2 commented:
Another one:

```
1 selection, buffer size 9474 startline 9618 endline - startline+1 -144

Thread 1 "AppRun" received signal SIGSEGV, Segmentation fault.
0x00000000008afd23 in TTextEdit::getSelectedText(QChar const&) ()
```

vadi2 commented:
```
1 selection, buffer size 9384 startline 9709 endline - startline+1 -325

Thread 1 "AppRun" received signal SIGSEGV, Segmentation fault.
0x00000000008afd23 in TTextEdit::getSelectedText(QChar const&) ()
```

SlySven committed:
Merge branch 'development' into BugFix_tryAndPreventCrashesWithHighlighting

SlySven committed:
Revise: combine two separate but related comments into one place

This was suggested during peer-review.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I have merged in changes that have been made into the `development` branch and combined that partial duplicate comment with the other one.

:thinking: 
> 1 selection, buffer size 9090 startline 9364 endline - startline+1 -274

and
>1 selection, buffer size 9384 startline 9709 endline - startline+1 -325

Is it possible that that crash is after the buffer has been shrunk down because it has reached the limit? For `startline` to be so far past the `buffer size` looks unsafe...!

vadi2 commented:
Yeah, seems to be the case?

Edru2 commented on src/TTextEdit.cpp:
```diff
@@ -1428,23 +1427,26 @@ QString TTextEdit::getSelectedText(const QChar& newlineChar)
     // mPB QPoint where selection ended
 
```
```suggestion
    // try to prevent crash if buffer is batch deleted
    if (mPA.y() > mpBuffer->lineBuffer.size() - 1 || mPB.y() > mpBuffer->lineBuffer.size() - 1){
        mPA.ry() -= mpBuffer->mBatchDeleteSize;
        mPB.ry() -= mpBuffer->mBatchDeleteSize;
    }
```

No idea if this is a good or bad solution (I'm sure you know) but it prevents crashes as I described at https://github.com/Mudlet/Mudlet/issues/3999#issuecomment-692311438 without altering the copied content.

vadi2 committed:
Try Edru's fix

Co-authored-by: Manuel Wegmann <60551052+Edru2@users.noreply.github.com>

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -1425,7 +1425,11 @@ QString TTextEdit::getSelectedText(const QChar& newlineChar)
     // mPB QPoint where selection ended
-
+    // try to prevent crash if buffer is batch deleted
+    if (mPA.y() > mpBuffer->lineBuffer.size() - 1 || mPB.y() > mpBuffer->lineBuffer.size() - 1){
```
Might I suggest replacing the `if (` with a `while (` as, if the buffer size is really small, and there is a *lot* of long lines coming in, doing this once *might* not be enough...

Edru2 replied:
sure :+1: 
I think you are more experienced with this part of code and know what is going on here. 
For me it was more shooting in the dark and it seemed to work but we need surely look out for other issues this could create.

SlySven replied:
I did, in the past try to insert an event generator that would fire every time the `TBuffer::shrinkbuffer(...)` method was called for any `TConsole` so that user scripts/packages that kept references (indexes) to line numbers could have a chance to know that they needed to remove a corresponding number of lines from those indexes into a particular console if it did get shrunk. However, IIRC it didn't work as expected and got dropped - I think it might be useful to resurrect it if possible. Does anyone else have any thoughts about that?

SlySven commented:
One thing - are we absolutely certain we can not have things, such as timer driven code or other things that might be inserting text into the `TBuffer` class asynchronously. It is something that I have wondered about in the past and I think we worked out that although Mudlet is running as a multiple thread application only the main (GUI) thread was changing things in that area of the codebase - but if we have got that wrong then we are well into ***[nasal daemon](https://en.wikipedia.org/wiki/Undefined_behavior)*** territory!

vadi2 commented:
Yes, that is certain. Would have definitely have had issues in that area of we did... But we haven't so far.

vadi2 committed:
Remove qDebug()s
